### DISCUSSION TITLE
Emit Event() && Revert()

### DISCUSSION THREAD

**Discussion by 0xPublicGoods** [Cyfrin/foundry-full-course-cu#72] [Q&A] [answered]
Hey Team

Im on the DEFI Lesson and was wondering about when to emit events, in the code we are emitting the event after the state change to ```s_CollateralDeposited``` and then if the collateral transfer fails, we revert, which will revert the state change.

My question is, will this revert the event already emitted? Should we rather emit the event after a successful call to the collateral token contract for transfer? this way it only emits if we have a successful transfer?

```solidity
    /*
     * @notice Follows CEI (Checks, Effects, Interaction) Pattern
     * @param tokenCollateralAddress The address of the collateral token
     * @param amountCollateral The amount of collateral to deposit
     */
    function depositCollateral(address tokenCollateralAddress, uint256 amountCollateral)
        external
        moreThanZero(amountCollateral)
        isAllowedToken(tokenCollateralAddress)
        nonReentrant
    {
        s_collateralDeposited[msg.sender][tokenCollateralAddress] += amountCollateral;

        // should this emit not go after the success check?
        emit CollateralDeposited(msg.sender, tokenCollateralAddress, amountCollateral);

        bool success = IERC20(tokenCollateralAddress).transferFrom(msg.sender, address(this), amountCollateral);

        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }
```

**Comment by cromewar** [comment, marked as answer]
Hello @0xPublicGoods 
What a nice question, well let me answer that the best I can.
1. When to trigger events? Each time you have a function which changes the state of the blockchain, this could be the change of variable, adding a new struct, or anything that make any type of changes on the blockchain. Do we have to emit a event for any banal change on any variable? Well is not 100% a must, but a good practice indeed.

2. What happens if the event gets emitted but the transaction revert? well the state changes are rolled back and discarded, including any events emitted during the execution of the transaction. In other words, if a transaction gets reverted, any events that were emitted during the execution of the transaction will also be reverted.
